### PR TITLE
feat: Win/Linux Dual Compatible `WORKSPACE` + Upgrade CUDA + Upgrade PyT

### PR DESCRIPTION
**Pull request by gs-olive**
# Description
- Add support for OS compatibility in the `WORKSPACE` file and `BUILD` files across the repository
- Upgrade bazel version to enable Windows builds in MSVC command prompt
- Validate multiple builds on CI, including with new Bazel version
- Upgrade default to CUDA 12.4
- Upgrade PyTorch versions

## Steps to Build on Windows Using the New `WORKSPACE`
Portions adapted from discussion with @HolyWu in #2806 
- Clone Torch-TRT (this repo)
- Open x64 Native Tools Command Prompt for VS 2022
- Install latest version of Torch (i.e. with `pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu121`)
- Run `set DISTUTILS_USE_SDK=1`
- This is analogous to setting `LD_LIBRARY_PATH` on Linux
- Run `python setup.py bdist_wheel`
- Run `python -m pip install dist/*.whl`

## Type of change
- New feature (non-breaking change which adds functionality)

# Checklist:

- [ x ] My code follows the style guidelines of this project (You can use the linters)
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ x ] I have made corresponding changes to the documentation
- [ x ] I have added tests to verify my fix or my feature
- [ x ] New and existing unit tests pass locally with my changes
- [ x ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Comment by gs-olive** [comment]
Installation instruction update

File: WORKSPACE [hunk 1]
```diff
@@ -46,6 +46,12 @@ new_local_repository(
     path = "/usr/local/cuda-12.1/",
 )
 
+new_local_repository(
+    name = "cuda_win",
+    build_file = "@//third_party/cuda:BUILD",
+    path = "C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.1/",
```
Comment (narendasan) [inline WORKSPACE:52]: Can you bump to 12.4 for both linux and windows?

File: docsrc/getting_started/getting_started_with_windows.rst [hunk 2]
```diff
@@ -7,133 +7,28 @@ Torch-TensorRT has community support for Windows platform using CMake
...
-Build configuration
+Build steps
 -------------------
 
-* Open Microsoft Visual Studio
-* Open Torch-TensorRT source code folder
-* Open Manage configurations -> Edit JSON to open CMakeSettings.json file.
-* Configure the CMake build configurations. Following is an example configuration:
-
-.. code-block:: none
```
Comment (gs-olive) [inline docsrc/getting_started/getting_started_with_windows.rst:24]: Is it useful to leave the old Visual Studio instructions in this tutorial?

**Review by peri044 — APPROVED** [review]
LGTM

File: docsrc/getting_started/getting_started_with_windows.rst (hunk 2 above)
Comment (peri044) [reply docsrc/getting_started/getting_started_with_windows.rst:24]: The existing windows support was using cmake right ?

File: core/conversion/BUILD [hunk 3]
```diff
@@ -26,10 +33,10 @@ cc_library(
         "//core/conversion/var",
         "//core/ir",
         "//core/util:prelude",
-        "@tensorrt//:nvinfer",
-    ] + select({
-        ":use_pre_cxx11_abi": ["@libtorch_pre_cxx11_abi//:libtorch"],
-        "//conditions:default": ["@libtorch"],
+        ] + select({
```
Comment (HolyWu) [inline core/conversion/BUILD:36]: unexpected extra indentation

File: core/conversion/conversionctx/BUILD [hunk 4]
```diff
@@ -21,10 +28,10 @@ cc_library(
     deps = [
         "//core/ir",
         "//core/util:prelude",
-        "@tensorrt//:nvinfer",
-    ] + select({
-        ":use_pre_cxx11_abi": ["@libtorch_pre_cxx11_abi//:libtorch"],
-        "//conditions:default": ["@libtorch"],
+        ] + select({
```
Comment (HolyWu) [inline core/conversion/conversionctx/BUILD:31]: unexpected extra indentation

File: docsrc/getting_started/getting_started_with_windows.rst [hunk 5]
```diff
@@ -7,133 +7,28 @@ Torch-TensorRT has community support for Windows platform using CMake
...
-
-    - Clean, configure, build
-
-e.g. /.vscode/settings.json
+* Open the app "x64 Native Tools Command Prompt for VS 2022" - note that Admin priveleges may be necessary
+* Ensure Bazelisk (Bazel launcher) is installed on your machine and available from the command line. Package installers such as Chocolatey can be used to install Bazelisk
+* Install latest version of Torch (i.e. with `pip install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu124`)
+* Clone the Torch-TensorRT repository and navigate to its root directory
+* Run `pip install ninja`
+* Run `pip install -r requirements-dev.txt`
```
Comment (HolyWu) [inline docsrc/getting_started/getting_started_with_windows.rst:23]: The packages in `requirements-dev.txt` seem to be more suitable for developers who write or debug codes. For end users who simply want to build the wheel, the essential packages to install are just `numpy, PyYAML, setuptools, torch, wheel`.

File: third_party/tensorrt/archive/BUILD [hunk 6]
```diff
@@ -160,26 +186,21 @@ cc_library(
...
-    name = "nvinferplugin",
-    visibility = ["//visibility:public"],
+    linkopts = ["-lpthread"],
     deps = [
         "nvinfer",
-        "nvinferplugin_headers",
-        "nvinferplugin_lib",
-        "@cuda//:cudart",
-    ],
+        ] + select({
```
Comment (HolyWu) [inline third_party/tensorrt/archive/BUILD:201]: unexpected extra indentation

File: third_party/tensorrt/local/BUILD [hunk 7]
```diff
@@ -103,10 +103,9 @@ cc_library(
     deps = [
         "nvinfer_headers",
         "nvinfer_lib",
-        "@cuda//:cudart",
-    ] + select({
-        ":windows": ["nvinfer_static_lib"],
-        "//conditions:default": [],
+        ] + select({
```
Comment (HolyWu) [inline third_party/tensorrt/local/BUILD:106]: unexpected extra indentation

File: third_party/tensorrt/local/BUILD [hunk 8]
```diff
@@ -359,7 +358,9 @@ cc_library(
     }),
     deps = [
         "nvinfer",
-        "@cuda//:cudart",
-    ],
+        ] + select({
```
Comment (HolyWu) [inline third_party/tensorrt/local/BUILD:361]: unexpected extra indentation

File: dev_dep_versions.yml [hunk 9]
```diff
@@ -1,3 +1,3 @@
-__version__: "2.4.0.dev0"
-__cuda_version__: "12.1"
-__tensorrt_version__: "10.0.1"
\ No newline at end of file
+__version__: "2.5.0.dev0"
+__cuda_version__: "12.4"
+__tensorrt_version__: "10.0.1"
```
Comment (gs-olive) [inline dev_dep_versions.yml:3]: Should the version here be increased to `2.5.0` and should the `version.txt` be modified similarly?
https://github.com/pytorch/TensorRT/blob/643fb7b20dcac74c27da428e8bb3e631a906be25/version.txt#L1
cc: @narendasan

**Pull request merged**